registrations/signals.py

```diff
@@ -148,6 +149,42 @@ def registration_post_save(
         transaction.on_commit(lambda: _recalculate_registration_capacities(instance.pk))
 
 
+@receiver(
+    pre_save,
+    sender=RegistrationUserAccess,
+    dispatch_uid="send_invitation_on_registration_user_access_pre_save",
+)
+def send_invitation_on_registration_user_access_pre_save(
+    sender: type[RegistrationUserAccess],
+    instance: RegistrationUserAccess,
+    **kwargs: dict,
+) -> None:
+    if not instance.pk:
+        return
+
+    old_instance = RegistrationUserAccess.objects.filter(pk=instance.pk).first()
+    if old_instance and (
+        old_instance.email != instance.email
+        or old_instance.is_substitute_user != instance.is_substitute_user
+    ):
+        instance.send_invitation()
```

Comment:
The signals code here is overall fine, but I was thinking that since `pre_save` is called before actually saving, the actual saving might fail after the invitation has been sent. Perhaps both the `pre_save` and `post_save` could be done by overriding the `Model.save()` method:
```
def save(self, *args, **kwargs):
    if self.pk:
        old_instance = RegistrationUserAccess.objects.filter(pk=self.pk).first()
    else:
        old_istance = None
         
    super().save(*args, **kwargs)
     
    if not old_instance or old_instance.email != self.email or old_instance.is_substitute_user != self.is_substitute_user:
        self.send_invitation()
```

Reply:
Changed as suggested and added `@transaction.atomic` for the `save` method